File: backend/src/models/Ambulance.js
const { Model, DataTypes } = require('sequelize');

class Ambulance extends Model {
  static initModel(sequelize) {
    Ambulance.init({

  id: {
    type: DataTypes.INTEGER,
    primaryKey: true,
    autoIncrement: true
  },
  providerName: {
    type: DataTypes.STRING,
    allowNull: false
  },
  phone: {
    type: DataTypes.STRING,
    allowNull: false
  },
  email: {
    type: DataTypes.STRING,
    allowNull: true
  },
  address: {
    type: DataTypes.TEXT,
    allowNull: false
  },
  city: {
    type: DataTypes.STRING,
    allowNull: false
  },
  state: {
    type: DataTypes.STRING,
    allowNull: false
  },
  zipCode: {
    type: DataTypes.STRING,
    allowNull: false
  },
  latitude: {
    type: DataTypes.DECIMAL(10, 8),
    allowNull: true
  },
  longitude: {
    type: DataTypes.DECIMAL(11, 8),
    allowNull: true
  },
  vehicleNumber: {
    type: DataTypes.STRING,
    allowNull: false
  },
  capacity: {
    type: DataTypes.INTEGER,
    allowNull: false,
    defaultValue: 1
  },
  ambulanceType: {
    type: DataTypes.ENUM('basic', 'advanced', 'cardiac', 'neonatal', 'air'),
    defaultValue: 'basic'
  },
  status: {
    type: DataTypes.ENUM('available', 'assigned', 'busy', 'maintenance'),
    defaultValue: 'available'
  },
  isAvailable: {
    type: DataTypes.BOOLEAN,
    defaultValue: true
  },
  responseTime: {
    type: DataTypes.INTEGER, // in minutes
    allowNull: true
  },
  rating: {
    type: DataTypes.DECIMAL(3, 2),
    allowNull: true
  },
  isVerified: {
    type: DataTypes.BOOLEAN,
    defaultValue: false
  },
  contactPerson: {
    type: DataTypes.STRING,
    allowNull: true
  },
  emergencyNumber: {
    type: DataTypes.STRING,
    allowNull: false
  }

    }, {
      sequelize,
      modelName: 'Ambulance',

  timestamps: true,
  tableName: 'ambulances',
  indexes: [
    {
      fields: ['city', 'state']
    },
    {
      fields: ['isAvailable']
    },
    {
      fields: ['status']
    },
    {
      fields: ['ambulanceType']
    }
  ]

    });
    return Ambulance;
  }

  static associate(models) {
    // Associations will be defined in models/index.js
  }
}

module.exports = Ambulance;

File: fastapi_backend/models/ambulance.py
from beanie import Document
from pydantic import BaseModel, Field
from typing import Optional
from datetime import datetime
from enum import Enum
from decimal import Decimal

class AmbulanceType(str, Enum):
    BASIC = "basic"
    ADVANCED = "advanced"
    CARDIAC = "cardiac"
    NEONATAL = "neonatal"
    AIR = "air"

class AmbulanceStatus(str, Enum):
    AVAILABLE = "available"
    ASSIGNED = "assigned"
    BUSY = "busy"
    MAINTENANCE = "maintenance"

# Pydantic schemas
class AmbulanceBase(BaseModel):
    provider_name: str
    phone: str
    email: Optional[str] = None
    address: str
    city: str
    state: str
    zip_code: str
    latitude: Optional[Decimal] = None
    longitude: Optional[Decimal] = None
    vehicle_number: str
    capacity: int = Field(default=1, ge=1)
    ambulance_type: AmbulanceType = AmbulanceType.BASIC
    status: AmbulanceStatus = AmbulanceStatus.AVAILABLE
    is_available: bool = True
    response_time: Optional[int] = None  # in minutes
    rating: Optional[Decimal] = Field(None, ge=0, le=5)
    is_verified: bool = False
    contact_person: Optional[str] = None
    emergency_number: str

class AmbulanceCreate(AmbulanceBase):
    pass

class AmbulanceUpdate(BaseModel):
    provider_name: Optional[str] = None
    phone: Optional[str] = None
    email: Optional[str] = None
    address: Optional[str] = None
    city: Optional[str] = None
    state: Optional[str] = None
    zip_code: Optional[str] = None
    latitude: Optional[Decimal] = None
    longitude: Optional[Decimal] = None
    vehicle_number: Optional[str] = None
    capacity: Optional[int] = Field(None, ge=1)
    ambulance_type: Optional[AmbulanceType] = None
    status: Optional[AmbulanceStatus] = None
    is_available: Optional[bool] = None
    response_time: Optional[int] = None
    rating: Optional[Decimal] = Field(None, ge=0, le=5)
    is_verified: Optional[bool] = None
    contact_person: Optional[str] = None
    emergency_number: Optional[str] = None

class AmbulanceResponse(AmbulanceBase):
    id: str
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True

# Beanie Document
class Ambulance(Document, AmbulanceBase):
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)

    class Settings:
        name = "ambulances"
        indexes = [
            "city",
            "state",
            "is_available",
            "status",
            "ambulance_type",
            "is_verified",
            "created_at"
        ]

    def __str__(self):
        return f"Ambulance(id={self.id}, provider_name={self.provider_name}, city={self.city})"
